### DISCUSSION TITLE
What's the proper way to set Foreground on CheckBox using Lightweight styling?

### DISCUSSION THREAD

**Discussion by jeanplevesque** [unoplatform/Uno.Themes#1214] [Q&A] [answered]
With the [Lightweight styling introduced with Uno 4.10](https://platform.uno/blog/4-10-release/) I can no longer set the `Foreground` property on CheckBox.
Instead I need to override the following 18 resources (I didn't include the variants for the disables states because I don't need them).
I want to set the `OnPrimaryBrush` to all 18.

```
CheckBoxForegroundUnchecked
CheckBoxForegroundUncheckedPointerOver
CheckBoxForegroundUncheckedPressed

CheckBoxForegroundChecked
CheckBoxForegroundCheckedPointerOver
CheckBoxForegroundCheckedPressed

CheckBoxForegroundIndeterminate
CheckBoxForegroundIndeterminatePointerOver
CheckBoxForegroundIndeterminatePressed

CheckBoxGlyphForegroundUnchecked
CheckBoxGlyphForegroundUncheckedPointerOver
CheckBoxGlyphForegroundUncheckedPressed

CheckBoxGlyphForegroundChecked
CheckBoxGlyphForegroundCheckedPointerOver
CheckBoxGlyphForegroundCheckedPressed

CheckBoxGlyphForegroundIndeterminate
CheckBoxGlyphForegroundIndeterminatePointerOver
CheckBoxGlyphForegroundIndeterminatePressed
```

I'd like to avoid writing 18 lines of code in my pages that need this. I assume I can do this using a style however I don't understand how to do that.
The [current documentation](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/lightweight-styling.html) doesn't show such use case and seems to redefine brushes with new colors.
In my case I want to reuse the existing `OnPrimaryBrush`.

**Comment by kazo0** [comment]
Good catch, CheckBox is a good example to highlight in the docs as it is a bit of a special case where Foreground/Background don't apply as you might expect, and definitely something that we missed when talking about breaking changes.

If you are attempting to use the Foreground property to manage the color of the checkmark/indeterminate glyph, you'll have to override the resources like you would if you were using the [Fluent version of the CheckBox style](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.checkbox?view=windows-app-sdk-1.4#control-style-and-template).

![image](https://github.com/unoplatform/uno/assets/4793020/82c7f6f3-3d1c-4370-bfd4-053dac9a2315)

See a similar question [here](https://stackoverflow.com/questions/51034396/how-to-change-checkboxs-background-for-uwp)

In the case of using an existing brush, your XAML should look something like:

```xml
<CheckBox Content="Checked">
	<CheckBox.Resources>
		<StaticResource x:Key="CheckBoxGlyphForegroundChecked" ResourceKey="OnPrimaryBrush" />
		<StaticResource x:Key="CheckBoxGlyphForegroundCheckedPointerOver" ResourceKey="OnPrimaryBrush" />
		<StaticResource x:Key="CheckBoxGlyphForegroundCheckedPressed" ResourceKey="OnPrimaryBrush" />
		<StaticResource x:Key="CheckBoxGlyphForegroundChecked" ResourceKey="OnPrimaryBrush" />
		<!-- ... -->
	</CheckBox.Resources>
</CheckBox>
```
 
Although, in the case of the Glyph Foreground, those [should already be all](https://github.com/unoplatform/Uno.Themes/blob/d19569b466a8dc4e533d5fcedee113804592d75e/src/library/Uno.Material/Styles/Controls/v2/CheckBox.xaml#L114-L122) OnPrimaryBrush

It's a bit verbose, I know, but we're trying to follow the standards from Fluent styles. There is work underway to be able to define a set of Resources as an attached property that you can use as a Setter on a custom Style so you can re-use the same CheckBox style with all the overridden resources throughout your app. If this is something you want to be done globally or at a Page-level, you can define these StaticResource aliases in the App/Page Resources instead of on the specific CheckBox.

There's also work starting around a simpler way to override the material colors at any scope. So you will be able to say "just for this CheckBox, use X as PrimaryColor/OnPrimaryColor/SurfaceColor/etc."

You are correct that the documentation should show a case like this where you can alias existing brushes as part of your lightweight styling overrides.

**Reply by jeanplevesque** [reply]
I'm not just interested in the glyph but also the `Content` (text) of the `Checkbox`.

1. Is this going to work with **themes**? I see that you use `StaticResource`, but `OnPrimaryBrush` should be referenced as a `ThemeResource` I believe?

1. Do you have an example that shows the override more globally (like as a style or as an app-wide override)?

**Reply by jeanplevesque** [reply]
I tried the following (with `ThemeResource`) but it doesn't work. (The overrides are not applied.)
![image](https://github.com/unoplatform/Uno.Themes/assets/39710855/a88846bd-7625-409d-92c2-da10ed7db043)
With `StaticResource` the overrides are applied, but it doesn't adjust when changing theme.

**Reply by kazo0** [reply]
Yeah the naming is confusing but `<StaticResource x:Key="..." ResourceKey="..." />` is the proper syntax for defining aliases. There's no such thing as a `<ThemeResource x:Key="..." ResourceKey="..." />` If you want them to be themeable you need to define them in themedictionaries like here: https://github.com/unoplatform/Uno.Themes/blob/d19569b466a8dc4e533d5fcedee113804592d75e/doc/lightweight-styling.md?plain=1#L27-L38
https://github.com/unoplatform/Uno.Themes/blob/d19569b466a8dc4e533d5fcedee113804592d75e/doc/lightweight-styling.md?plain=1#L45

Except use the StaticResource alias and not the new SolidColorBrush definition

More info on aliasing here: https://learn.microsoft.com/en-us/windows/apps/design/style/xaml-styles#aliasing-control-resources-to-new-names

**Reply by kazo0** [reply]
As for a more global way, it would be like is described [here](https://platform.uno/docs/articles/external/uno.themes/doc/lightweight-styling.html#apppage-level-styling)

You just move the the ResourceDictionary containing your ThemeDictionaries to the Page.Resources or the AppResources

**Reply by jeanplevesque** [reply]
Ok but in that example the brush is being redefined with custom colors. In my case I don't want to do that. I want to leverage the existing `OnPrimaryBrush` that is already themed.

**Reply by kazo0** [reply]
Unfortunately it doesn't matter if it's already a theme resource, when aliasing you must define the alias for each theme if you want it to work across all themes

<img width="448" alt="image" src="https://github.com/unoplatform/Uno.Themes/assets/4793020/940bb328-ffa3-4b5e-b0b8-6dccd1ca7ec3">

**Reply by jeanplevesque** [reply]
Ok so I need 36 overrides: 18 per theme (duplicated).

In my page I have a `ListView` that contain CheckBoxes so if I override using `Page.Resource` it changes them all and that's not what I want... I need to override a specific one.
Ideally I would like to avoid adding 36 lines on that one `CheckBox` and hide that in a `Style`. How can I do that?

**Reply by Xiaoy312** [reply]
you can create a OneSpecialButton.xaml with the resources, and in Button.Resources import that with Source=path\to\OneSpecialButton.xaml

**Reply by Xiaoy312** [reply]
this can also be converted into a style too, but you will need to create a custom attached-property to set .Resources in code behind by the assigned value, as FrameworkElement.Resources is not a dp. (i havent really tested this, but in theory this hsould work)

**Reply by kazo0** [reply]
Right now you can't hide it in a Style, unless you re-template and remove the fact that the visual states set different brushes on the ContentPresenter for the Foreground. That's where what I mentioned before with the work to have an attached property for the Resources that can be set in a Style.Setter will come in handy 

You should only need to override these 9 brushes:

> CheckBoxForegroundUnchecked
> CheckBoxForegroundUncheckedPointerOver
> CheckBoxForegroundUncheckedPressed
> 
> CheckBoxForegroundChecked
> CheckBoxForegroundCheckedPointerOver
> CheckBoxForegroundCheckedPressed
> 
> CheckBoxForegroundIndeterminate
> CheckBoxForegroundIndeterminatePointerOver
> CheckBoxForegroundIndeterminatePressed

As the Glyph ones are already set to OnPrimaryBrush

Basically your options are the same as you would have from this answer: https://stackoverflow.com/questions/60950009/checkbox-content-color-cant-be-set-in-uwp

You either override the brushes in the resources or you have a custom CheckBox style that doesn't change certain properties per visual state.

You _might_ get away with doing something like 

```xml
<CheckBox.Resources>
	<SolidColorBrush x:Key="CheckBoxForegroundUnchecked" Color={ThemeResource OnPrimaryColor} />
	<!-- ... -->
</CheckBox.Resource> 
```

But I'm not sure if that works in all cases

**Reply by jeanplevesque** [reply, marked as answer]
Ok so your suggestion of redefining a new `SolidColorBrush` using a `ThemeResource` reference on the color is what I ended up doing because it avoids duplicating the overrides per theme.
Thanks for that.

```xml
<CheckBox x:Name="TermsOfServiceCheckBox"
	Content="I agree to the terms of services"
	x:Uid="CreateAccount_TermsOfServices"
	IsChecked="{Binding Form.AgreeToTermsOfServices, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}"
	HorizontalAlignment="Left">
	<CheckBox.Resources>
		<SolidColorBrush x:Key="CheckBoxForegroundUnchecked" Color="{ThemeResource OnPrimaryColor}" />
		<SolidColorBrush x:Key="CheckBoxForegroundUncheckedPointerOver" Color="{ThemeResource OnPrimaryColor}" />
		<SolidColorBrush x:Key="CheckBoxForegroundUncheckedPressed" Color="{ThemeResource OnPrimaryColor}" />

		<SolidColorBrush x:Key="CheckBoxForegroundChecked" Color="{ThemeResource OnPrimaryColor}" />
		<SolidColorBrush x:Key="CheckBoxForegroundCheckedPointerOver" Color="{ThemeResource OnPrimaryColor}" />
		<SolidColorBrush x:Key="CheckBoxForegroundCheckedPressed" Color="{ThemeResource OnPrimaryColor}" />

		<SolidColorBrush x:Key="CheckBoxCheckBackgroundStrokeUnchecked" Color="{ThemeResource OnPrimaryColor}" />
		<SolidColorBrush x:Key="CheckBoxCheckBackgroundStrokeUncheckedPointerOver" Color="{ThemeResource OnPrimaryColor}" />
		<SolidColorBrush x:Key="CheckBoxCheckBackgroundStrokeUncheckedPressed" Color="{ThemeResource OnPrimaryColor}" />
	</CheckBox.Resources>
</CheckBox>
```

The code is pretty heavy though.
So like you mentioned, having recipes to make the code more lightweight would truly bring value.

Should I open issues about this? (Or perhaps upvote or comment existing issues?)

**Reply by kazo0** [reply]
@jeanplevesque yeah I agree this is a bit much for customizing, CheckBox is a bit of a special case with the way the template is built but there will still be other cases where this sort of thing is needed.

The issue for tracking a Resources attached property to be used in something like a Style.Setter is here: https://github.com/unoplatform/Uno.Themes/issues/1116, maybe this should be moved to Toolkit eventually

The issue for having a way to basically re-init the MateralTheme with a different color override is here: https://github.com/unoplatform/Uno.Themes/issues/938, this needs some more info but basically we want to achieve "scoped" color overrides where you can simply say that within a specific resource scope: primary/surface/etc. color should be X and all of the brushes will properly adjust to the new base color properly
